Joomla 4.0: Removing calls to JError::raiseError()

Hackwar opened the pull request:
This PR removes almost all calls to JError::raiseError() and replaces them with either a JViewGenericdataexception or a JUserAuthorizationException.

wilsonge commented on administrator/components/com_messages/controllers/config.php:
```diff
@@ -37,9 +37,7 @@ public function save()
-			JError::raiseError(500, $model->getError());
-
-			return false;
+			throw new Exception($model->getError(), 500);
```
This one in the controller means we can probably do something better than this?

wilsonge commented on administrator/components/com_users/controllers/groups.php:
```diff
@@ -51,8 +51,7 @@ public function delete()
 		{
-			JError::raiseError(500, JText::_('JERROR_ALERTNOAUTHOR'));
-			jexit();
+			throw new Exception(JText::_('JERROR_ALERTNOAUTHOR'), 500);
```
OK Let's make these 403's whilst we are here

wilsonge commented on administrator/components/com_users/controllers/groups.php:
```diff
@@ -71,8 +70,7 @@ public function publish()
 		{
-			JError::raiseError(500, JText::_('JERROR_ALERTNOAUTHOR'));
-			jexit();
+			throw new Exception(JText::_('JERROR_ALERTNOAUTHOR'), 500);
```
403

wilsonge commented on administrator/components/com_users/controllers/groups.php:
```diff
@@ -91,8 +89,7 @@ public function reorder()
 		{
-			JError::raiseError(500, JText::_('JERROR_ALERTNOAUTHOR'));
-			jexit();
+			throw new Exception(JText::_('JERROR_ALERTNOAUTHOR'), 500);
```
403

wilsonge commented on administrator/components/com_users/controllers/groups.php:
```diff
@@ -111,8 +108,7 @@ public function saveorder()
 		{
-			JError::raiseError(500, JText::_('JERROR_ALERTNOAUTHOR'));
-			jexit();
+			throw new Exception(JText::_('JERROR_ALERTNOAUTHOR'), 500);
```
403

wilsonge commented on administrator/components/com_users/controllers/groups.php:
```diff
@@ -131,8 +127,7 @@ public function checkin()
 		{
-			JError::raiseError(500, JText::_('JERROR_ALERTNOAUTHOR'));
-			jexit();
+			throw new Exception(JText::_('JERROR_ALERTNOAUTHOR'), 500);
```
403

wilsonge commented on administrator/components/com_users/controllers/level.php:
```diff
@@ -57,8 +57,7 @@ public function delete()
 		{
-			JError::raiseError(500, JText::_('JERROR_ALERTNOAUTHOR'));
-			jexit();
+			throw new Exception(JText::_('JERROR_ALERTNOAUTHOR'), 500);
```
403

wilsonge commented on administrator/modules/mod_latest/helper.php:
```diff
@@ -82,9 +82,7 @@ public static function getList(&$params)
-			JError::raiseError(500, $error);
-
-			return false;
+			throw new Exception(implode("\n", $errors), 500);
```
OK So whilst we are here should a module failing stop the entire page from loading??

mbabker replied:
This one should probably be caught by the module's base file and enqueue a message I'd say.  Modules probably shouldn't kill the entire page but it'll have to be a case-by-case to figure out what to do.

wilsonge commented on components/com_banners/models/banner.php:
```diff
@@ -52,7 +52,7 @@ public function click()
 		catch (RuntimeException $e)
 		{
-			JError::raiseError(500, $e->getMessage());
+			throw new Exception($e->getMessage(), 500);
```
Can we use `setError` here for now please and `return false` and let the view handle the error for consistency

wilsonge commented on components/com_mailto/controller.php:
```diff
@@ -96,7 +96,7 @@ public function send()
 				if (strpos($_POST[$field], $header) !== false)
 				{
-					JError::raiseError(403, '');
+					throw new Exception('', 403);
```
Let's define a message for the love of god

Hackwar commented:
I reverted several of the changes to concentrate on the view.*.php files. I also introduced 2 new exceptions, JViewGenericdataexception and JUserAuthorizationException. Those 2 are consistently used throughout the view.*.php files. I'm wondering if it wouldn't be better to create a new branch and new PR with just those view.*.php files...

Hackwar committed:
Removing large parts of JError from view.*.php files

mbabker commented:
There are also a couple new Exception classes added for either 3.6.3 or 3.7
from Andre.  I think his was only a controller exception but I say it
because I think it dealt with authorization.

On Thursday, October 6, 2016, Hannes Papenberg <notifications@github.com>
wrote:

> I reverted several of the changes to concentrate on the view.*.php files.
> I also introduced 2 new exceptions, JViewGenericdataexception and
> JUserAuthorizationException. Those 2 are consistently used throughout the
> view.*.php files. I'm wondering if it wouldn't be better to create a new
> branch and new PR with just those view.*.php files...
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/12272#issuecomment-252107430>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAWfodYFEqP2cgK9trq2_5jZ7Bpp8aTiks5qxXctgaJpZM4KMJdF>
> .
>

Hackwar commented:
I've reverted most of the stuff and started over and just concentrated on the view files. Taking babysteps here... I would keep it at this and do further refactoring in other PRs.

wilsonge commented on components/com_mailto/views/mailto/view.html.php:
```diff
@@ -58,9 +58,7 @@ protected function &getData()
-			JError::raiseError(403, JText::_('COM_MAILTO_LINK_IS_MISSING'));
-
-			return false;
+			throw new Exception(JText::_('COM_MAILTO_LINK_IS_MISSING'), 403);
```
I think this is a 400 not a 403?

wilsonge commented:
This is beginning to look significantly better :)

Hackwar committed:
Changing com_mailto exception from 403 to 400

Hackwar committed:
Codestyle

Hackwar committed:
Codestyle

Hackwar commented:
Should we merge this for the moment or do you want more changes before accepting this?

Hackwar committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into j4error

Hackwar committed:
Merge branch 'j4error' of github.com:Hackwar/joomla-cms into j4error
